### ISSUE TITLE
Handle failed signup when backend and frontend databases get out of sync

### ISSUE THREAD

**Issue by taoeffect** [okTurtles/group-income#495]
### Problem

During development, in between server runs, the database gets wiped. However, the frontend still keeps track of all contracts it was subscribed to. Then, when a dev tries to create a new user, the user successfully gets created, but the login fails because the frontend calls `syncContractWithServer` on a contract that no longer exists on the backend, resulting in a "500 Internal Server" error in the signup modal and console output that looks like this:

```
[Log] [sbp] okTurtles.events/emit (app.js, line 4544)
Array (2)
0 "open-modal"
1 {__ob__: Observer, …}

Array Prototype
[Log] [sbp] namespace/lookup – ["asdf"] (1) (app.js, line 4544)
[Error] Failed to load resource: the server responded with a status of 404 (Not Found) (asdf, line 0)
[Log] [sbp] gi/contract/create – ["IdentityContract", {attributes: {name: "asdf", email: "asdf@asdf.com", picture: "http://localhost:8000/simple/assets/images/default-avatar.png"}}] (2) (app.js, line 4544)
[Log] [sbp] gi/contract/create – ["MailboxContract", {}] (2) (app.js, line 4544)
[Log] [sbp] backend/publishLogEntry – [GIMessage] (1) (app.js, line 4544)
[GIMessage]Array (1)
[Log] [sbp] backend/publishLogEntry – [GIMessage] (1) (app.js, line 4544)
[GIMessage]Array (1)
[Log] [sbp] gi/contract/create-action – ["IdentitySetAttributes", {mailbox: "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6"}, "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"] (3) (app.js, line 4544)
[Log] [sbp] backend/latestHash – ["21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"] (1) (app.js, line 4544)
[Log] [sbp] backend/publishLogEntry – [GIMessage] (1) (app.js, line 4544)
[GIMessage]Array (1)
[Log] [sbp] namespace/register – ["asdf", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"] (2) (app.js, line 4544)
[Log] [sbp] backend/latestHash – ["21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"] (1) (app.js, line 4544)
[Log] syncContractWithServer(): 21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4 latestHash is: 21XWnNFP9Hd9bpxsSxNDY7KhivE3nBvARCxnQpayjnva7D7XrC (app.js, line 5870)
[Log] Now Synchronizing Contract: 21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4 its most recent was undefined but the latest is 21XWnNFP9Hd9bpxsSxNDY7KhivE3nBvARCxnQpayjnva7D7XrC (app.js, line 5886)
[Log] [sbp] backend/eventsSince – ["21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"] (2) (app.js, line 4544)
[Log] addLogEntry(): – "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4" – {version: 1, previousHEAD: null, contractID: null, …} (app.js, line 5243)
{version: 1, previousHEAD: null, contractID: null, nonce: 2470883, action: Object}Object
[Log] [sbp] okTurtles.events/emit – ["contractsModified", {add: "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4"}] (2) (app.js, line 4544)
[Log] [sbp] okTurtles.events/emit – ["21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage] (3) (app.js, line 4544)
["21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage]Array (3)
[Log] [sbp] okTurtles.events/emit – ["event-handled", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage] (3) (app.js, line 4544)
["event-handled", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage]Array (3)
[Log] addLogEntry(): – "21XWnNFP9Hd9bpxsSxNDY7KhivE3nBvARCxnQpayjnva7D7XrC" – {version: 1, previousHEAD: "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", contractID: "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", …} (app.js, line 5243)
{version: 1, previousHEAD: "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", contractID: "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", nonce: 9964298, action: {type: "IdentitySetAttributes", data: {mailbox: "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6"}}}Object
[Log] [Backend] subscribed 21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4: – {type: "success", data: {type: "sub", id: "N~5_OW9ioBerBCR~szcRO"}} (app.js, line 3823)
[Log] [sbp] okTurtles.events/emit – ["21XWnNFP9Hd9bpxsSxNDY7KhivE3nBvARCxnQpayjnva7D7XrC", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage] (3) (app.js, line 4544)
["21XWnNFP9Hd9bpxsSxNDY7KhivE3nBvARCxnQpayjnva7D7XrC", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage]Array (3)
[Log] [sbp] okTurtles.events/emit – ["event-handled", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage] (3) (app.js, line 4544)
["event-handled", "21XWnNF9RpT3oRri98nXxvPfwawkyfFa28AGmkWmdgsufo1fF4", GIMessage]Array (3)
[Log] [sbp] backend/latestHash – ["21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6"] (1) (app.js, line 4544)
[Log] syncContractWithServer(): 21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6 latestHash is: 21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6 (app.js, line 5870)
[Log] Now Synchronizing Contract: 21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6 its most recent was undefined but the latest is 21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6 (app.js, line 5886)
[Log] [sbp] backend/eventsSince – ["21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6"] (2) (app.js, line 4544)
[Log] addLogEntry(): – "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6" – {version: 1, previousHEAD: null, contractID: null, …} (app.js, line 5243)
{version: 1, previousHEAD: null, contractID: null, nonce: 1514421, action: {type: "MailboxContract", data: {}}}Object
[Log] [sbp] okTurtles.events/emit – ["contractsModified", {add: "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6"}] (2) (app.js, line 4544)
[Log] [sbp] okTurtles.events/emit – ["21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", GIMessage] (3) (app.js, line 4544)
["21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", GIMessage]Array (3)
[Log] [sbp] okTurtles.events/emit – ["event-handled", "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", GIMessage] (3) (app.js, line 4544)
["event-handled", "21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6", GIMessage]Array (3)
[Log] [Backend] subscribed 21XWnNHoE4i4us8Me2sppBF3QApbFJxxBkboYwWc3ks8tcwAK6: – {type: "success", data: {type: "sub", id: "N~5_OW9ioBerBCR~szcRO"}} (app.js, line 3823)
[Log] [sbp] backend/latestHash – ["21XWnNTk19kmEThtJTjSrV65t7saxJASM7BH7cffi1WNF8pJct"] (1) (app.js, line 4544)
[Error] Failed to load resource: the server responded with a status of 500 (Internal Server Error) (21XWnNTk19kmEThtJTjSrV65t7saxJASM7BH7cffi1WNF8pJct, line 0)
[Log] [sbp] okTurtles.events/emit – ["logout"] (1) (app.js, line 4544)
[Log] Error: 500: Internal Server Error — misc.js:5 (app.js, line 13959)
```

On the backend we get something like:

```
Error: contract 21XWnNTk19kmEThtJTjSrV65t7saxJASM7BH7cffi1WNF8pJct hash no latest hash!
```

### Solution

It's not clear what the correct solution to this is, however, the error that should be returned in this case from the backend should be a 404 error, not a 500 error.

And on the frontend, I guess we should just drop the contract (forget about it entirely), or skip attempting to sync it in the `login` action handler in `state.js`. Perhaps if we're in development mode we drop it, and in production mode we skip it?

EDIT: see also #505 — **Update Troubleshooting.md with how to deal with out-of-sync db**

**Comment by simonv3** [comment]
For dev purposes, is there a reason to wipe the database every time on restart? Is it possible to just keep it? (I'm not sure what's involved here, and very likely out of my element). Maybe making that an optional flag with grunt dev like `grunt dev --clean-db`?

**Comment by taoeffect** [comment]
> For dev purposes, is there a reason to wipe the database every time on restart?

Yes:

1. It doesn't take a long time to setup a group, and most days you should be able to spend an entire doing development without having to wipe the database
2. It helps make sure the developers to don't break something that was working fine by forcing them to run through everything
3. It makes testing during development more reliable and consistent
4. It makes sure the database is fresh and clean for all our tests that run via mocha
5. It forces us to gracefully handle error conditions like the one above

**Comment by simonv3** [comment]
> 1. It doesn't take a long time to setup a group, and most days you should be able to spend an entire doing development without having to wipe the database

This doesn't really apply to irregular contributors though right? It seems like it's an extra and kind of unnecessary contribution burden to me (and for them I might even consider seeding the database with existing users).

> 2. It helps make sure the developers to don't break something that was working fine by forcing them to run through everything

That's what I thought the automated e2e tests were for?

> 3. It makes testing during development more reliable and consistent

I'm not sure I get this - could you elaborate?

> 4. It makes sure the database is fresh and clean for all our tests that run via mocha

I haven't looked deeply in how group-income's test suite is set up beyond running it, but in my mind (but I haven't seen this work in Javascript to be honest, but it's fairly standard practice in, like, Ruby on Rails with Rspec) the database should be clean for each test anyway, or at least for the test suite. Otherwise your tests aren't really independent of each other, and it's my understanding - and experience - that you end up not really knowing what each test is testing.

> 5. It forces us to gracefully handle error conditions like the one above

Just to be clear - I'm definitely not saying "don't handle it" :stuck_out_tongue:. I'm coming more from a place where if I'm working on a ticket on Monday, and I shut down my computer till Friday for whatever reason, then I might have to spend a not insignificant time to get my local environment back to the location that I was testing. This is fine if you're paying me, but as a volunteer I'd be very quickly frustrated, especially if I'm fixing some obscure side-bug only found within a certain condition of my database.

I actually really like the idea of being able to quickly start the database over again, and in that case we'd need to be handling this, but I'm not convinced by the justification of "running through the entire set-up process every time you start working on this project is good for ensuring things don't break".

But this might not be the place for that conversation, and I'm coming from a project where we have a lot of data, and it'd be a nightmare if we had to reset our project every time we started our dev environment, so I guess that's what's influencing my perspective on this.

**Comment by taoeffect** [comment]
@simonv3 that's totally reasonable. Feel free to open a separate issue to make it simple to persist the debase during development.

However, another reason we don't have this implemented is because we don't have persisting the database at all implemented. 😆 

That's something else that needs to be done before we can persist the database for development. And yeah, if it really helps your development we can bump it up in priority. It only takes me about 15 seconds or so to run through the signup + group creation process, so it's not too bad. But yeah, I agree, you raise good points. 👍

**Comment by simonv3** [comment]
Haha, yeah, see, I think a lot of it is "I'm still not sure what's meant to be working or not", so if I ever am just wrong please tell me!

Edit: my experience so far has really been "it's quick to get up and running with this" and sure, let's get a persistent DB up first and then we can worry about the dev-experience around it :P

**Comment by taoeffect** [comment]
Note: this issue is linked with #505 — **Update Troubleshooting.md with how to deal with out-of-sync db**

**Comment by taoeffect** [comment]
I'm re-opening the issue, as it wasn't entirely fixed by #593.

#593 made it so that `frontend/main.js` would call `sbp('gi.db/settings/delete', user)` if we loaded up the app with a missing `SETTING_CURRENT_USER`. However, this doesn't fix the problem with SignUp.vue where if we try to sign up with a user that was previously signed up. We'll still get an error like this:

```
Error: contract 21XWnNJXXyLFgmYazMGknWcyCBcDXzzuMrJdssVHSgZ2wiMTXS hash no latest hash!
```

The solution is to make sure we call `'gi.db/settings/delete'` on the newly created user in SignUp.vue before logging in ... and just be careful.

I.e. before this code is called in SignUp.vue:

```js
          await this.$store.dispatch('login', {
            name: this.form.name,
            identityContractId: user.hash()
          })
```

I.e. something like:

```js
          const oldSettings = await sbp('gi.db/settings/load', this.form.name)
          if (oldSettings) {
            console.warn(`deleting settings for pre-existing user ${this.form.name}!`, oldSettings)
            await sbp('gi.db/settings/delete', this.form.name)
          }
```

Also, an `await` needs to be added to `main.js:56`.
